hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/KeyPrefixContainerCodec.java

```diff
@@ -50,6 +59,113 @@ public Class<KeyPrefixContainer> getTypeClass() {
+      buffer.put(KEY_DELIMITER_BUFFER.duplicate());
+      buffer.putLong(object.getContainerId());
+    }
+
+    return buffer;
+  }
+
+  @Override
+  public KeyPrefixContainer fromCodecBuffer(@Nonnull CodecBuffer buffer) throws CodecException {
+
+    final ByteBuffer byteBuffer = buffer.asReadOnlyByteBuffer();
+    final int totalLength = byteBuffer.remaining();
+    final int startPosition = byteBuffer.position();
+    final int delimiterLength = KEY_DELIMITER_BYTES.length;
+
+    // Check for two delimiters + two longs (version and containerId)
+    if (totalLength >= 2 * (delimiterLength + Long.BYTES)) {
+      // Extract keyPrefix (everything except last 2 delimiters + 2 longs)
+      int keyPrefixLength = totalLength - 2 * delimiterLength - 2 * Long.BYTES;
+      byteBuffer.position(startPosition);
+      String keyPrefix = decodeStringFromBuffer(byteBuffer, keyPrefixLength);
+
+      // Skip delimiter and read version
+      byteBuffer.position(startPosition + keyPrefixLength + delimiterLength);
+      long version = byteBuffer.getLong();
+
+      // Skip delimiter and read containerId
+      byteBuffer.position(startPosition + keyPrefixLength + delimiterLength + Long.BYTES + delimiterLength);
+      long containerId = byteBuffer.getLong();
+
+      return KeyPrefixContainer.get(keyPrefix, version, containerId);
+    }
+
+    // Check for one delimiter + one long
+    if (totalLength >= delimiterLength + Long.BYTES) {
+      // Extract keyPrefix (everything except last delimiter + long)
+      int keyPrefixLength = totalLength - delimiterLength - Long.BYTES;
+      byteBuffer.position(startPosition);
+      String keyPrefix = decodeStringFromBuffer(byteBuffer, keyPrefixLength);
+
+      // Skip delimiter and read the long value
+      byteBuffer.position(startPosition + keyPrefixLength + delimiterLength);
+      long longValue = byteBuffer.getLong();
+
+      // Based on encoding logic: if keyVersion != -1, it's encoded first, then containerId
+      // So if we have only one long value, it should be the keyVersion
+      return KeyPrefixContainer.get(keyPrefix, longValue, -1);
+    }
+
+    // If we reach here, the buffer contains only the key prefix (no delimiters found)
+    byteBuffer.position(startPosition);
+    String keyPrefix = decodeStringFromBuffer(byteBuffer, totalLength);
+    return KeyPrefixContainer.get(keyPrefix, -1, -1);
+  }
+
+  /**
+   * Decode string from ByteBuffer without copying to intermediate byte array.
+   * Uses CharsetDecoder for efficient decoding.
+   */
+  private String decodeStringFromBuffer(ByteBuffer buffer, int length) throws CodecException {
+    if (length == 0) {
+      return "";
+    }
+
+    try {
+      ByteBuffer slice = buffer.duplicate();
+      slice.limit(slice.position() + length);
+
+      buffer.position(buffer.position() + length);
+
+      CharsetDecoder decoder = UTF_8.newDecoder();
+      CharBuffer charBuffer = decoder.decode(slice);
+      return charBuffer.toString();
```

Comment:
The `UTF_8.newDecoder()` behavior is different from `new String(bytes, UTF_8)`.  So, let's use `new String(bytes, UTF_8)` as before:
```java
  private static String decodeStringFromBuffer(ByteBuffer buffer) {
    if (buffer.remaining() == 0) {
      return "";
    }

    final byte[] bytes;
    if (buffer.hasArray()) {
      bytes = buffer.array();
    } else {
      bytes = new byte[buffer.remaining()];
      buffer.get(bytes);
    }
    return new String(bytes, UTF_8);
  }
```